### ISSUE TITLE
Funnel Metadynamics exposed a bug with CustomCentroidBondForce calculations 

### ISSUE THREAD

**Issue by CoryKornowicz** [openmm/openmm#3535]
I will need to amend this issue tomorrow when I have more time to generate more tests (what tests/logs would you like to see?) but there was a bug that spuriously appeared when I was setting up a funnel metadynamics simulation using the BioSimSpace implementation as a template for the CustomForces and such. Using OpenMM 7.7.0 from conda-forge and cudatoolkit == 11.2.2, there is a bug concerning CVForce calculation across multiple GPU's. I knew that it was an issue in BioSimSpace because they use openmm==7.4.2, but I did not think it carried this far down with all of the GPU updates. 

The suspected sections under question pertain to 
```
current_cvs = np.array(list(meta.getCollectiveVariables(simulation)) + [meta.getHillHeight(simulation)])
```
where "meta" is a Metadynamics object and the "getCollectiveVariables(simulation)" function does not seem to be communicating between the multiple contexts. 

I've hunted down this function 
https://github.com/openmm/openmm/blob/a5e42f57289b654368ba91183cb8f4a3649a5f68/platforms/cuda/include/CudaKernels.h#L360
and it seems to cast the CUDA kernel into a Common kernel. Is there any reason that it wouldn't be possible to sync the contexts together? This problem goes a little above my knowledge on where to begin.

**Comment by peastman** [comment]
It's not clear to me exactly what problem you're describing.  What exactly are you doing, what exactly happens when you do it, and how does it differ from what you expect to happen?  The more detail you can give, the better.

**Comment by CoryKornowicz** [comment]
I am performing a Funnel Metadynamics simulation. Unfortunately, I am unable to share my data files, however, the problem is agnostic of system configuration except for CudaDevices and a CustomCVForce. The example from the BioSimSpaceTutorial repo: [Funnel-MD](https://github.com/michellab/BioSimSpaceTutorials/tree/main/02_funnel_metad) works as a test set. 

If the device indexes here are change from any single value to multiple cards,  
https://github.com/michellab/BioSimSpaceTutorials/blob/1433371403d48f3b4aec8e83f8135cae59a2d089/02_funnel_metad/fun-metaD-work_dir/openmm.py#L104
`platform = Platform.getPlatformByName('CUDA')`
`properties = {'CudaDeviceIndex': '0'}`

(This file also has the definition of the CustomCVForce for the funnel and the Metadynamics system.)

the CustomCVForce has no way of collecting the right CV values since the CUDA kernel (and CalcCustomCVForceKernel by extension) is extended from CommonCalcCustomCVForceKernel. This implementation does not take into account that all of the contexts split between the CUDA devices are not synced before the positions are returned for the CV calculation. I was wondering if there was any way to amend the kernel, or would it dramatically slow down the speed of the simulation? 

Normally, on the single GPU implementation there is no difference between the CPU kernel and works as expected.

**Comment by peastman** [comment]
Do you have a test case that shows the problem?  If you can provide a simple (preferably self contained) way to reproduce it, that will make it a lot easier to debug.

**Comment by CoryKornowicz** [comment]
https://github.com/CoryKornowicz/Funnel_MD_Test

I have assembled a small example based on the BioSimSpaceTutorial and documented the difference in collective variable measurements when more than one device is used.

**Comment by jmichel80** [comment]
tagging @dlukauskis and @lohedges to follow this thread

**Comment by peastman** [comment]
I figured out the problem.  It isn't actually related to the CustomCVForce.  You can get the same thing by adding a CustomCentroidBondForce directly to the System.  The problem is that CustomCentroidBondForce isn't parallelized.  It runs everything on a single GPU.  As a result, it doesn't cooperate properly with multi-GPU parallelism to ensure everything is synchronized correctly.  It's possible to get a race condition where the force gets computed before the energy buffer is cleared at the start of the step.

There are a few possible ways of fixing it.  I'll run some tests to figure out which is best.

**Comment by peastman** [comment]
The fix is in #3560.
